spec/cli.js

```diff
@@ -40,4 +57,149 @@ describe('noflo-nodejs CLI', () => {
+        runtimeClient = c;
+        return c.connect();
+      }));
+    describe('setting component sources', () => {
+      const source = `const noflo = require('noflo');
+exports.getComponent = () => {
+  const c = new noflo.Component();
+  c.inPorts.add('in');
+  c.outPorts.add('out');
+  c.process((input, output) => {
+    output.sendDone(input.getData() + 2);
+  });
+  return c;
+};`;
+      const componentPath = path.resolve(__dirname, './fixtures/auto-save/components/Plusser.js');
+      after('clean up file', () => unlink(componentPath));
+      it('should be possible to send the source code to the runtime', () => runtimeClient
+        .protocol.component.source({
+          name: 'Plusser',
+          library: 'auto-save',
+          language: 'javascript',
+          code: source,
+        }));
+      it('should have saved the source code to the fixture folder', () => readFile(
+        componentPath,
+        'utf-8',
+      )
+        .then((contents) => {
+          expect(contents).to.eql(source);
+        }));
+    });
+    describe('setting component sources outside of project', () => {
+      let source;
+      const componentPath = path.resolve(__dirname, './fixtures/auto-save/components/Output.js');
+      before('read source code', () => readFile(
+        path.resolve(__dirname, '../node_modules/noflo-core/components/Output.js'),
+        'utf-8',
+      )
+        .then((contents) => {
+          source = contents;
+        }));
+      it('should be possible to send the source code to the runtime', () => runtimeClient
+        .protocol.component.source({
+          name: 'Output',
+          library: 'core',
+          language: 'javascript',
+          code: source,
+        }));
+      it('should not have saved the source code to the fixture folder', () => readFile(
+        componentPath,
+        'utf-8',
+      )
+        .then(
+          () => Promise.reject(new Error('core/Output was saved unexpectedly')),
+          () => Promise.resolve('No Output.js found, as expected'),
+        ));
+    });
+    describe('editing a graph without namespaced name', () => {
+      const graphName = 'Test';
+      const graphPath = path.resolve(__dirname, `./fixtures/auto-save/graphs/${graphName}.json`);
+      const graphInstance = new fbpGraph.Graph(graphName);
+      before('set up graph', () => {
+        graphInstance.setProperties({
+          ...graphInstance.properties,
+          library: 'auto-save',
+        });
+        graphInstance.addNode('one', 'auto-save/Plusser');
+        graphInstance.addNode('two', 'core/Output');
+        graphInstance.addEdge('one', 'out', 'two', 'in');
+        graphInstance.addInitial(1, 'one', 'in');
+      });
+      after('clean up file', () => unlink(graphPath));
+      it('should be possible to send a graph to the runtime', () => runtimeClient
+        .protocol.graph.send(graphInstance, false));
+      it('should have saved the graph JSON to the fixture folder', () => waitFor(100)
```

Comment:
Maybe an alternative to this could be subscribing `done` of the previous cases for an auto-save event. If the auto-save event makes sense anywhere outside the tests, of course.

Reply:
in these tests we're exercising a black box (separate process), so no access to events